sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/namedExpressions.scala

```diff
@@ -429,6 +429,39 @@ case class OuterReference(e: NamedExpression)
   final override val nodePatterns: Seq[TreePattern] = Seq(OUTER_REFERENCE)
 }
 
+/**
+ * A place holder used to hold a reference that has been resolved to a field outside of the current
+ * plan. This is used for correlated subqueries.
+ */
+case class VariableReference(varName: String, literal: Literal, canFold: Boolean = true)
+  extends LeafExpression {
+
+  override def dataType: DataType = literal.dataType
+  override def nullable: Boolean = literal.nullable
+  override def prettyName: String = "session." + varName
+  override def foldable: Boolean = canFold
+
+  override def sql: String = varName
+
+  final override val nodePatterns: Seq[TreePattern] = Seq(RUNTIME_REPLACEABLE)
```

Comment:
it's not a RuntimeReplaceable, we should add a new node pattern, or leave it as empty.

Reply:
It's not?

Reply:
OK, how about: VARIABLE_REFERENCE?

Reply:
SGTM. `RUNTIME_REPLACEABLE` should only be used for expressions extending `RuntimeReplaceable`

Reply:
But do we still want the same logic? I.e., I will trigger this new tag in replaceExpression()s...

Reply:
Done